src/main/kotlin/app/RepoHasher.kt

```diff
@@ -3,83 +3,199 @@
+
+    private fun loadGit(): Git? {
+        return try {
+            Git.open(File(localRepo.path))
+        } catch (e: IOException) {
+            Logger.error("Cannot access repository at path "
+                    + "$localRepo.path", e)
+            null
+        }
+    }
+
+    private fun closeGit() {
+        gitRepo?.close()
+        git?.close()
+    }
+
+    private fun getRepoRehashes() {
+        val initialRevCommit = getObservableRevCommits().blockingLast()
+        val initialCommit = Commit(initialRevCommit)
+        repo.initialCommitRehash = initialCommit.rehash
+
+        var repoRehash = repo.initialCommitRehash
+        if (localRepo.remoteOrigin.isNotBlank()) {
+            repoRehash += localRepo.remoteOrigin
+        } else {
+            repoRehash += localRepo.path + localRepo.userName
+        }
+
+        repo.rehash = DigestUtils.sha256Hex(repoRehash)
+    }
 
-        var firstCommit: RevCommit
+    private fun getRepoConfig() {
+        gitRepo ?: return
+        val config = gitRepo.getConfig()
+        localRepo.author = Author(name = config.getString("user",
+                                                          null,
+                                                          "name") ?: "",
+                                  email = config.getString("user",
+                                                           null,
+                                                           "email") ?: "")
+        localRepo.remoteOrigin = config.getString("remote",
+                                                  "origin",
+                                                  "url") ?: ""
+        localRepo.userName = try { System.getProperty("user.name") }
+                             catch (e: Exception) { "" }
+    }
 
-        // Get commits loading observable and wait for subscribers.
-        val observableRevCommits = getObservableRevCommits().publish()
+    private fun isKnownRepo(): Boolean {
+        if (Configurator.getRepos().find { it.rehash == repo.rehash } != null) {
+            return true
+        }
+        return false
+    }
 
-        // Main data flow. Read commits from repo and send them to server.
-        observableRevCommits.doOnNext {
-                    Logger.info("Commit: ${it.id.name} ${it.shortMessage}")
+    private fun findCommitsOverlap() {
+        addedCommits = mutableListOf()
+        removedCommits = repo.commits.toMutableList()
+        val lastServerCommit: Commit = repo.commits.last()
+        var isLastServerCommitChecked: Boolean = false
+
+        val commitsObservable = getObservableRevCommits()
+        commitsObservable.map({ Commit(it) }).filter {
+            val email = it.author.email
+            localRepo.hashAllAuthors || (email == localRepo.author.email ||
+                    repo.emails.contains(email))
+        }.blockingSubscribe({  // OnNext
+            Logger.info("Commit: ${it.rehash}")
+
+            if (!isLastServerCommitChecked) {
```

Comment:
With RxJava we can easily deal with asynchronous tasks we will use this really soon.